### PR TITLE
Print dates and UUIDs with tagged literals

### LINKED ISSUE #9: `#uuid` and `#inst` printing
Hi there, I noticed that UUIDs and `java.util.Date`s print as objects rather than as their tagged literal form.

```clojure
(ddiff/pretty-print (ddiff/diff #inst "2019-04-01T14:38:55.998-00:00"
                                #inst "2019-04-01T14:39:11.732-00:00"))
```
prints as
```diff
-#<java.util.Date@5c6429fc Mon Apr 01 10:38:55 EDT 2019>
+#<java.util.Date@69ad72db Mon Apr 01 10:39:11 EDT 2019>
```
and
```clojure
(ddiff/pretty-print (ddiff/diff #uuid "70fa1988-a1d0-42f3-9c81-8d60ac3e999a"
                                #uuid "0f2386f7-0587-46c7-8211-b3919f07fca8"))
```
prints as
```diff
-#<java.util.UUID@68e1706a 70fa1988-a1d0-42f3-9c81-8d60ac3e999a>
+#<java.util.UUID@5a4ce851 0f2386f7-0587-46c7-8211-b3919f07fca8>
```

would you be interested in a patch that includes printers for these?

### PR DESCRIPTION
**Pull request by nwjsmith**
Fixes #9

### CODE REVIEW

File: src/lambdaisland/deep_diff/printer.clj [hunk 1]
```diff
@@ -34,6 +36,16 @@
      [:align (interpose [:span (:map-delimiter this) :line] entries)]
      (color/document this :delimiter "}")]))
 
+(def ^:private thread-local-utc-date-format
+  (proxy [ThreadLocal] []
```
Comment (plexus) [inline src/lambdaisland/deep_diff/printer.clj]: Today I learned `SimpleDateFormat` isn't thread safe! Thanks for catching this.

File: test/lambdaisland/deep_diff/printer_test.clj [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+(defn- printed
+  [diff]
+  (let [printer (printer/puget-printer {})]
+    (with-out-str (-> diff
+                      (printer/format-doc printer)
+                      (printer/print-doc printer)))))
+
+(deftest print-doc-test
+  (testing "date"
+    (is (= "[31m-#inst \"2019-04-09T14:57:46.128-00:00\"[0m [32m+#inst \"2019-04-10T14:57:46.128-00:00\"[0m\n"
```
Comment (plexus) [inline test/lambdaisland/deep_diff/printer_test.clj]: I would use `\u001b` here instead of a literal escape character, makes it a bit easier to edit.

**Comment by plexus** [comment]
Thanks, good stuff! I added a tiny nitpick, would you mind changing that? otherwise I'm happy to merge.

**Comment by nwjsmith** [comment]
@plexus Good catch, added in the character literals to improve the readability. It looks like Clojure [prints `java.util.Calendar`s and `java.sql.Timestamp`s as `#inst` literals too](https://github.com/clojure/clojure/blob/master/src/clj/clojure/instant.clj#L188-L234). Should I add print handlers for those?

**Comment by plexus** [comment]
That'd be great. I'm a bit surprised about `Calendar`, as it's really a bit of a separate category. `java.sql.Timestamp` is a subclass of `Date` so that makes sense, but anyway if that's what Clojure's printer does I guess we can do the same.

**Comment by nwjsmith** [comment]
@plexus all set

**Comment by plexus** [comment]
Nice! Will put out a release of deep-diff and kaocha shortly.

**Comment by plexus** [comment]
Oops, spoke too soon. Seems the generative tests found a problem: https://circleci.com/gh/lambdaisland/deep-diff/65?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

```
FAIL in lambdaisland.deep-diff.diff-test/round-trip-diff (diff_test.clj:207)
expected: {:result true}
  actual: {:shrunk {:total-nodes-visited 96, :depth 13, :pass? false, :result false, :result-data nil, :time-shrinking-ms 31, :smallest [{##NaN 0} {}]}, :failed-after-ms 90, :num-tests 40, :seed 1554909672622, :fail [{.L-gZA6*FF "\tqñIkÉ»y?äbÍà", ##NaN -0.06219482421875, \| 34/23, true \8, 1 #uuid "a2565b7c-d09e-45de-890d-efd8de6c6cc0", :_-/b!3?*?x !*eUdRI/*.*sl?:8} {#uuid "528f95c8-8ab7-4547-b1ad-c7059f45a3f5" :?4, *! -26/15, 354756707 9751, :x0_6_0wR_ -37224235}], :result false, :result-data nil, :failing-size 39, :pass? false, :test-var "round-trip-diff"}
```

I can't say for sure if this is related to your change. Would you have time to check out what's happening here?

**Comment by plexus** [comment]
Looking at the output I guess it's due to `##NaN`, which isn't your problem. I'm also fine with merging this and making a separate issue for that.

**Pull request merged**